docs/proposals/THV-2591-operator-registry-integration-via-gateway-api.md

```diff
@@ -0,0 +1,473 @@
+2. **`toolhive.stacklok.dev/registry-url`** (on MCP resource): Direct URL for non-Gateway API users
+3. **`toolhive.stacklok.dev/registry-name`** (on either): Override registry entry name
+
+**Metadata Annotations** (on HTTPRoute or MCP resource):
+4. **`toolhive.stacklok.dev/registry-tier`**: Server classification ("Official", "Community", "Partner")
+5. **`toolhive.stacklok.dev/registry-tools`**: Comma-separated list of tool names (e.g., "create_pr,merge_pr,list_issues")
+6. **`toolhive.stacklok.dev/registry-tags`**: Comma-separated categorization tags (e.g., "github,vcs,production")
+7. **`toolhive.stacklok.dev/registry-description`**: Human-readable description (overrides default)
+
+### Discovery Tracking
+
+The operator tracks discovered endpoints using Kubernetes Events for observability rather than extending MCPRegistry status. This keeps the MCPRegistry CRD focused on registry management concerns.
+
+**Event Types:**
+- `EndpointDiscovered`: When a new endpoint is discovered via HTTPRoute or annotation
+- `EndpointUpdated`: When an endpoint URL changes (Gateway address change, HTTPRoute update)
+- `EndpointRemoved`: When HTTPRoute annotation removed or resource deleted
+
+**Event Example:**
+```yaml
+apiVersion: v1
+kind: Event
+metadata:
+  name: github-mcp-discovered
+  namespace: mcp-servers
+type: Normal
+reason: EndpointDiscovered
+message: "Discovered endpoint for MCPServer/github-mcp via HTTPRoute/github-mcp-external: https://mcp.company.com/github"
+involvedObject:
+  apiVersion: toolhive.stacklok.dev/v1alpha1
+  kind: MCPRegistry
+  name: company-registry
+```
+
+This approach provides:
+- Clear audit trail of discovery events
+- No CRD status bloat
+- Standard Kubernetes observability pattern
+- Easy integration with monitoring systems
+
+### Registry Entry Format
+
+Discovered endpoints are published using the upstream MCP Registry format (from modelcontextprotocol/registry) with ToolHive publisher extensions:
+
+```json
+{
+  "servers": [
+    {
+      "server": {
+        "$schema": "https://static.modelcontextprotocol.io/schemas/2025-10-17/server.schema.json",
+        "name": "com.company.mcp/github-production",
+        "description": "GitHub MCP server for production use",
+        "version": "1.0.0",
+        "remotes": [
+          {
+            "type": "sse",
+            "url": "https://mcp.company.com/github"
+          }
+        ],
+        "_meta": {
+          "io.modelcontextprotocol.registry/publisher-provided": {
+            "io.github.stacklok": {
+              "https://mcp.company.com/github": {
+                "status": "active",
+                "tier": "Official",
+                "tools": ["create_pr", "merge_pr", "list_issues"],
+                "tags": ["github", "vcs", "production"]
+              }
+            }
+          }
+        }
+      },
+      "_meta": {
+        "io.modelcontextprotocol.registry/official": {
```

Comment:
That part is probably obsolete as this metadata is supposed to come if we were to get an MCP server from the upstream registry.